vault/logical_raw.go

```diff
@@ -97,6 +115,10 @@ func (b *RawBackend) handleRawRead(ctx context.Context, req *logical.Request, da
 // handleRawWrite is used to write directly to the barrier
 func (b *RawBackend) handleRawWrite(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
 	path := data.Get("path").(string)
+	compressed := data.Get("compressed").(bool)
+	compressionType := data.Get("compression_type").(string)
```

Comment:
Hmm, I understand why you added `compression_type`: if we want to allow arbitrary writes, including writes to create storage entries that don't presently exist, it makes sense to provide this option.

However, similar to my original concern that users could get the compression wrong if we make them do it themselves, I have the same concern here: most users aren't going to know what compression type is appropriate to a given storage path, and I'd rather they not have to study the code to find out, if we can avoid it.  For the case where the write is going to update an existing value - which I presume is going be by far the most common case - can we use compressutil to work out what the compression type of the current stored value is, and thereby make the param optional for updates?

Reply:
There is no function in compressutil to return the type of compression used.   I could write one (but my understanding is it would take 3 diffs (one for compressutil in the SDK, one to update the go.mod to reference the new sdk, and then this diff).

Auto-detection behavior could be confusing (only works for updates, but not new values).  And I'm not sure specifying compression_type is that bad.  Because you can read the existing value (without decompression), a user can find the canary byte to know what an existing compression was.

But I'm up for doing multiple diffs if that's what's required to push this diff through.

Reply:
> my understanding is it would take 3 diffs (one for compressutil in the SDK, one to update the go.mod to reference the new sdk, and then this diff

We have a `replace => ./sdk` directive in go.mod to use the sdk within the repo, the actual version of the sdk in Vault's go.mod isn't really used.  So you should be able to include the change in this one PR.

> Auto-detection behavior could be confusing (only works for updates, but not new values).

Yeah, I'm not thrilled about that either, but I've never seen sys/raw used to create new values, only delete, read, and update.  I think "creates" should be possible but I'm okay with them being harder (and with this inconsistency) if it makes things better for the much more common update case. 

> And I'm not sure specifying compression_type is that bad. Because you can read the existing value (without decompression), a user can find the canary byte to know what an existing compression was.

It's not that bad, agreed, but please bear in mind that the primary use case we have for sys/raw is: a customer is down, Vault won't start, and this is the only way to get them running again.  In such a high-stress urgent context, anything we can do to make things easier is desirable.

Reply:
Thanks for the pointers.  I've updated write so it will infer compression type from the existing value (if it exists and was compressed) -- error out if it cannot infer.  Users can still specify it for new values.

Reply:
@ncabatoff -- Do you have any other feedback on this change?

Reply:
Yes, but not in this thread.